contracts/0.8.9/StakingRouter.sol

```diff
@@ -0,0 +1,570 @@
+        uint256 _stakingModuleIndex,
+        uint256 _totalDepositsCount
+    ) internal view returns (uint256) {
+        (, uint256[] memory newKeysAllocation, StakingModuleCache[] memory modulesCache) = _getKeysAllocation(_totalDepositsCount);
+        return newKeysAllocation[_stakingModuleIndex] - modulesCache[_stakingModuleIndex].activeKeysCount;
+    }
+
+    /**
+     * @notice return shares table
+     *
+     * @return recipients recipients list
+     * @return moduleFees fee of each recipient
+     * @return totalFee total fee to mint for each module and treasury
+     */
+    function getStakingRewardsDistribution()
+        external
+        view
+        returns (address[] memory recipients, uint16[] memory moduleFees, uint16 totalFee)
+    {
+        (uint256 totalActiveKeys, StakingModuleCache[] memory modulesCache) = _loadNotStoppedStakingModulesCache();
+        uint256 modulesCount = modulesCache.length;
+
+        /// @dev return empty response if there are no modules or active keys
+        if (modulesCount == 0 || totalActiveKeys == 0) {
+            return (new address[](0), new uint16[](0), 0);
+        }
+
+        recipients = new address[](modulesCount);
+        moduleFees = new uint16[](modulesCount);
+
+        uint256 rewardedModulesCount = 0;
+        uint256 moduleKeysShare;
+        for (uint256 i; i < modulesCount; ++i) {
+            /// @dev skip modules which have no active keys
+            if (modulesCache[i].activeKeysCount > 0) {
+                moduleKeysShare = ((modulesCache[i].activeKeysCount * TOTAL_BASIS_POINTS) / totalActiveKeys);
+
+                recipients[i] = address(modulesCache[i].stakingModuleAddress);
+                moduleFees[i] = uint16((moduleKeysShare * modulesCache[i].moduleFee) / TOTAL_BASIS_POINTS);
+
+                totalFee += uint16((moduleKeysShare * modulesCache[i].treasuryFee) / TOTAL_BASIS_POINTS) + moduleFees[i];
+                unchecked {
+                    rewardedModulesCount++;
+                }
+            }
+        }
+
+        /// @dev shrink arrays
+        if (rewardedModulesCount < modulesCount) {
+            uint256 trim = modulesCount - rewardedModulesCount;
+            assembly {
+                mstore(recipients, sub(mload(recipients), trim))
+                mstore(moduleFees, sub(mload(moduleFees), trim))
+            }
+        }
+    }
+
+    /// @notice returns new deposits allocation after the distribution of the `_keysToAllocate` keys
+    function getKeysAllocation(uint256 _keysToAllocate) public view returns (uint256 allocated, uint256[] memory allocations) {
+        (allocated, allocations, ) = _getKeysAllocation(_keysToAllocate);
+    }
+
+    /**
+     * @dev Invokes a deposit call to the official Deposit contract
+     * @param _maxDepositsCount max deposits count
+     * @param _stakingModuleId id of the staking module to be deposited
+     * @param _depositCalldata module calldata
+     */
+    function deposit(
```

Comment:
Add event about Deposit - as an antipod to `StakingRouterETHReceived`

Reply:
added